sky/packages/sky/lib/src/gestures/double_tap.dart

```diff
@@ -0,0 +1,164 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'arena.dart';
+import 'constants.dart';
+import 'events.dart';
+import 'recognizer.dart';
+import 'tap.dart';
+
+class DoubleTapGestureRecognizer extends DisposableArenaMember {
+  static int sInstances = 0;
+
+  DoubleTapGestureRecognizer({ this.router, this.onDoubleTap });
+
+  // Implementation notes:
+  // The double tap recognizer can be in one of four states. There's no
+  // explicit enum for the states, because they are already captured by
+  // the state of existing fields.  Specifically:
+  // Waiting on first tap: In this state, the _trackers list is empty, and
+  // _firstTap is null.
+  // First tap in progress: In this state, the _trackers list contains all
+  // the states for taps that have begun but not completed. This list can
+  // have more than one entry if two pointers begin to tap.
+  // Waiting on second tap: In this state, one of the in-progress taps has
+  // completed successfully. The _trackers list is again empty, and
+  // _firstTap records the successful tap.
+  // Second tap in progress: Much like the "first tap in progress" state, but
+  // _firstTap is non-null.  If a tap completes successfully while in this
+  // state, the callback is invoked and the state is reset.
+  // There are various other scenarios that cause the state to reset:
+  // - All in-progress taps are rejected (by time, distance, pointercancel, etc)
+  // - The long timer between taps expires
+  // - The gesture arena decides we have been rejected wholesale
+
+  PointerRouter router;
+  GestureTapCallback onDoubleTap;
+
+  Timer _doubleTapTimer;
+  TapTracker _firstTap;
+  Map<int, TapTracker> _trackers = new Map<int, TapTracker>();
+
+  void addPointer(PointerInputEvent event) {
+    // Ignore out-of-bounds second taps
+    if (_firstTap != null &&
+        !_firstTap.isWithinTolerance(event, kDoubleTapTouchSlop))
+      return;
+    _stopDoubleTapTimer();
+    TapTracker tracker = new TapTracker(
+      event: event,
+      entry: GestureArena.instance.add(event.pointer, this)
+    );
+    _trackers[event.pointer] = tracker;
+    tracker.startTimer(() => _reject(tracker));
+    tracker.startTrackingPointer(router, handleEvent);
+  }
+
+  void handleEvent(PointerInputEvent event) {
+    TapTracker tracker = _trackers[event.pointer];
+    assert(tracker != null);
+    if (event.type == 'pointerup') {
+      if (_firstTap == null)
+        _registerFirstTap(tracker);
+      else
+        _registerSecondTap(tracker);
+    } else if (event.type == 'pointermove' &&
+        !tracker.isWithinTolerance(event, kTouchSlop)) {
+      _reject(tracker);
+    } else if (event.type == 'pointercancel') {
+      _reject(tracker);
+    }
+  }
+
+  void acceptGesture(int pointer) {}
```

Comment:
Yeah, it's a little odd how the class hierarchy ended up.  I pulled dispose() out of GestureRecognizer so that the GestureDetector could call it on all known recognizers.  Other than that and acceptGesture() I didn't need anything out of GestureRecognizer, and its assumptions about how pointer tracking worked didn't quite jibe with what I was doing, so I dropped it from the inheritance.